biochemist-python/chapters/File_Parsing.ipynb

```diff
@@ -0,0 +1,602 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "After you open and read information from a file object, you should always close the file."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "outfile.close()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## An alternative way to open a file.\n",
+    "Alternatively, you can open a file using context-manager. In this case, the context manager will automatically handle closing of the file. To use a context manager to open and close the file, you use the word with, and put everything you want to be done while the file is open in an indented block."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"exercise-this-is-a-title exercise\">\n",
+    "<p class=\"exercise-title\">Check your understanding</p>\n",
+    "    <p>Check that your file was read in correctly by determining how many lines are in the file.</p>\n",
+    "    \n",
+    "```{admonition} Answer\n",
+    ":class: dropdown\n",
+    "\n",
+    "```python\n",
+    "print(len(data)) \n",
+    "    \n",
+    "2232\n",
+    "```\n",
+    "</div>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Searching for a pattern in your file\n",
+    "The file we opened is the complete PDB file for the [Crystal structure of multidrug-resistant clinical isolate 769 HIV-1 protease in complex with ritonavir](https://www.rcsb.org/structure/4EYR).  As stated previously, the `readlines()` function put the file contents into a list where each element is a line of the file. You may remember from lesson 1 that a `for` loop can be used to execute the same code repeatedly. As we learned in the previous lesson, we can use a for loop to iterate through elements in a list.\n",
+    "\n",
+    "Let's take a look at what's in the file.\n",
+    "\n",
+    "```python\n",
+    "for line in data:\n",
+    "    print(line)\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This will print exactly what is in the file.\n",
+    "\n",
+    "If you look through the output, you will find a line that starts with HETNAM. This line contains information about the ligand or small molecule that is bound to HIV protease in this structure, providing both the abbreviation (RIT) and the full name of the drug (RITONAVIR).  We want to search through this file and find that line, and print only that line. We can do this using an `if` statement.\n",
+    "\n",
+    "Returning to our file example,"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for line in data:\n",
+    "    if 'HETNAM' in line:\n",
```

Comment:
does this happen several times? (as in, lines with 'HETNAM' appears several times?). In that case, many lines will be printed, but only the last one will be stored. Is there any line which will occur only once?

Reply:
In 4eyr.pdb, there is only one HETNAM line, representing the single non-water heterogroup in the structure, so this works with 4eyr.pdb. In a file such as 2w9g.pdb, there are three lines that start with HETNAM, representing three non-water heterogroups in the structure, so the three lines would print, but only the last one would be stored as HETNAM_line. I can’t think of a line that is guaranteed to occur only once in a PDB file. This is actually a flaw in the PDB file format and one of the reasons that the PDB has switched to cif format for their files. I think the example here serves the purpose of the notebook. If someone wants to dig deeper, they will encounter the problem you mentioned. Should I include a note on how to change the loop if there is more than one HETNAM line?

Reply:
Since there can be more than one `HETNAM`, do you see any value in keeping a list of them? Perhaps this is a case where you would want to create an empty list outside of the loop, and append the `HETATM` lines. This could even be an exercise - "more than one `HETNAM` line can occur in a pdb file, can you think of a way to keep all of the lines using syntax we have already learned?"

Reply:
DONE. The output is not elegant, but it contains all of the heterogen names and abbreviations as a list.